architecture-docs/adr-002.md

```diff
@@ -0,0 +1,490 @@
+# ADR 002: Finalize specifications for interaction between Validator Set Changes, Punishments and Rewards
+
+## Changelog
+- 23-01-2020: Initial Draft
+
+## Context
+
+Currently, `ValidatorState` is used by punishments and rewards as a source of data to compute their respective slashing
+ratio and rewards amount. Besides this, `ValidatorState` is also used to manage dynamic validator set (i.e.,
+validators can be added or removed based on their network operations).
+
+### Problems with current design
+
+1. Our current design does not delete validator information from `ValidatorState` when a validator either unbonds their
+   funds or gets removed from validator set because of some punishment. The reason for not deleting this information is
+   because it is still required for computing `SlashRatio` while slashing a punished validator. Because of this,
+   validators cannot rejoin validator set (because their information already exists and we cannot add a duplicate
+   entry).
+1. During rewards calculation, we calculate `total_stake` based of the active validator at the time of rewards
+   distribution. This does not take into account all the validator set changes that took place in current rewards
+   period. For example, when a validator unbonds or deposits funds just before rewards distribution (let us assume that
+   they unbonded their funds at 23rd hour), their stake is not considered in rewards calculation.
+
+Besides this, validator set, rewards pool and punishments are three different components of our system and should
+maintain their own data instead of relying on `ValidatorState` for information. There should be a specified
+interface/behavior for interaction between these components.
+
+## Decision
+
+This document proposes new design for Dynamic Validator Set, Punishments and Rewards Pool. At the end, it specifies the
+interaction between all three components. 
+
+### Dynamic Validator Set
+
+`Validator` struct holds all the information related to a validator/council node.
+
+```rust
+/// A validator in Crypto.com Chain (this data structure can be merged with existing `CouncilNode`)
+pub struct Validator {
+    /// Name of validator
+    pub name: String,
+    /// Staking address of validator's staking account
+    pub staking_address: StakedStateAddress,
+    /// Tendermint's public key of validator
+    pub tendermint_pub_key: TendermintPubKey,
+    /// Current voting power of validator
+    pub voting_power: TendermintVotePower,
```

Comment:
Does this field guarantee the stored value either greater than minimal voting power or zero,
Or we do the minimal voting power filtering at query time?

Reply:
While writing this ADR, I assumed that whoever implements it will put all the necessary bound checks.